### PR TITLE
opentelemetry: set status code

### PR DESCRIPTION
**Pull request by zirain**
Commit Message: set span status code base on http status code for OpenTelemetry tracer, xref: https://github.com/istio/istio/issues/50195
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #32114
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/tracers/opentelemetry/tracer.cc [hunk 1]
```diff
@@ -120,7 +121,24 @@ void Span::setAttribute(absl::string_view name, const OTelAttribute& attribute_v
   *span_.add_attributes() = key_value;
 }
 
-void Span::setTag(absl::string_view name, absl::string_view value) { setAttribute(name, value); }
+void Span::setTag(absl::string_view name, absl::string_view value) {
+  if (name == Tracing::Tags::get().HttpStatusCode) {
+    uint64_t status_code;
+    // For HTTP status codes in the 5xx range, as well as any other code the client failed to
```
Comment (silverstar194) [inline source/extensions/tracers/opentelemetry/tracer.cc:127]: Can you check the formatting here? The comment looks like it has strange line breaks.

Comment (zirain) [reply source/extensions/tracers/opentelemetry/tracer.cc:127]: reformated by `tools/local_fix_format.sh`

File: test/extensions/tracers/opentelemetry/opentelemetry_tracer_impl_test.cc [hunk 2]
```diff
@@ -559,6 +564,7 @@ TEST_F(OpenTelemetryDriverTest, ExportOTLPSpanWithAttributes) {
   span->setTag("", "empty_tag_value");
   // Overwrite a tag.
   span->setTag("first_tag_name", "first_tag_new_value");
+  span->setTag("http.status_code", "500");
```
Comment (adisuissa) [inline test/extensions/tracers/opentelemetry/opentelemetry_tracer_impl_test.cc:567]: Consider adding this in its own explicit test (avoiding a test covering multiple use-cases).

File: source/extensions/tracers/opentelemetry/tracer.cc [hunk 3]
```diff
@@ -120,7 +121,25 @@ void Span::setAttribute(absl::string_view name, const OTelAttribute& attribute_v
...
+    if (absl::SimpleAtoi(value, &status_code)) {
+      if (status_code >= 500 ||
+          (status_code >= 400 &&
+           span_.kind() == ::opentelemetry::proto::trace::v1::Span::SPAN_KIND_CLIENT)) {
+        span_.mutable_status()->set_code(
+            ::opentelemetry::proto::trace::v1::Status::STATUS_CODE_ERROR);
+      }
+    }
+  }
+  setAttribute(name, value);
```
Comment (adisuissa) [inline source/extensions/tracers/opentelemetry/tracer.cc:141]: Just to clarify, is it intentional to have both the attribute and the status code?
I see the benefit when introducing a new feature, just want to better understand whether this is desired.

**Review by adisuissa — COMMENTED** [review]
Overall LGTM.
Please add a release note.
Assigning @htuch as codeowner.
/assign @htuch

File: source/extensions/tracers/opentelemetry/tracer.cc (hunk 3 above)
Comment (zirain) [reply source/extensions/tracers/opentelemetry/tracer.cc:141]: I confirmed with a lot of imeplements in OTel-collector, most of them set both.

File: changelogs/current.yaml [hunk 4]
```diff
@@ -10,6 +10,9 @@ behavior_changes:
 
 minor_behavior_changes:
 # *Changes that may cause incompatibilities for some users, but should not for most*
+- area: tracers
+  change: |
+    Fix status code always unset for OpenTelemetry tracers.
```
Comment (htuch) [inline changelogs/current.yaml:15]:
```suggestion
    Set status code for OpenTelemetry tracers (previously unset).
```

**Review by htuch — DISMISSED** [review]
LGTM modulo nit comment.

**Review by wbpcode — APPROVED** [review]

**Pull request merged**